psi4/src/psi4/libfock/composite.cc

```diff
@@ -0,0 +1,796 @@
+    }
+
+    // Temporary buffers for J to minimize race tonditions
+    std::vector<std::vector<SharedMatrix>> JT;
+
+    for (int thread = 0; thread < nthread_; thread++) {
+        std::vector<SharedVector> gp2;
+        std::vector<SharedMatrix> J2;
+        for (size_t ind = 0; ind < nmat; ind++) {
+            J2.push_back(std::make_shared<Matrix>(max_nbf_per_shell, max_nbf_per_shell));
+            J2[ind]->zero();
+        }
+        JT.push_back(J2);
+    }
+
+    // Solve for gammaP = (P|rs)*Drs
+#pragma omp parallel for num_threads(nthread_) schedule(dynamic)
+    for (int P = 0; P < aux_nshell; P++) {
+
+        int p_start = auxiliary_->shell_to_basis_function(P);
+        int num_p = auxiliary_->shell(P).nfunction();
+
+        int thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+
+        for (const auto& UV : shell_pairs) {
+
+            int U = UV.first;
+            int V = UV.second;
+
+            int u_start = primary_->shell_to_basis_function(U);
+            int num_u = primary_->shell(U).nfunction();
+
+            int v_start = primary_->shell_to_basis_function(V);
+            int num_v = primary_->shell(V).nfunction();
+
+            ints_[thread]->compute_shell(P, 0, U, V);
+
+            const double* buffer = ints_[thread]->buffer();
+
+            double prefactor = 2.0;
+            if (U == V) prefactor *= 0.5;
+
+            for (int i = 0; i < D.size(); i++) {
+                double** Dp = D[i]->pointer();
+                double *Puv = const_cast<double *>(buffer);
+
+                /*
+                for (int p = p_start; p < p_start + num_p; p++) {
+                    int dp = p - p_start;
+                    for (int u = u_start; u < u_start + num_u; u++) {
+                        int du = u - u_start;
+                        for (int v = v_start; v < v_start + num_v; v++) {
+                            int dv = v - v_start;
+                            gamp[i * naux + p] += prefactor * (*Puv) * Dp[u][v];
+                            Puv++;
+                        }
+                    }
+                }
+                */
+
+                std::vector<double> Dbuff(num_u * num_v, 0.0);
+                for (int u = u_start; u < u_start + num_u; u++) {
+                    int du = u - u_start;
+                    for (int v = v_start; v < v_start + num_v; v++) {
+                        int dv = v - v_start;
+                        Dbuff[du * num_v + dv] = Dp[u][v];
+                    }
+                }
+                C_DGEMV('N', num_p, num_u * num_v, prefactor, (double *)Puv, num_u * num_v, Dbuff.data(), 1, 1.0, &(gamp[i * naux + p_start]), 1);
+            }
+        }
+    }
+
+    // Solve for gammaQ, (P|Q) * gammaQ = gammaP
+    SharedMatrix Jmet_copy = Jmet_->clone();
+
+    std::vector<int> ipiv(naux);
+    C_DGESV(naux, nmat, Jmet_copy->pointer()[0], naux, ipiv.data(), gamp, naux);
+
+#pragma omp parallel for num_threads(nthread_) schedule(dynamic)
```

Comment:
Two points regarding the parallelism in this algorithm:
1. The first loop, the one forming gammaP, parallelizes over auxiliary shells, while this second loop, the one forming J, parallelizes over one of the regular shell indices u. Why are they different?
2. If rearranging how the loops are set up isn't a big problem (likely depending on the answer to point #1), it might be more efficient from a parallel perspective to parallelize over the combined regular shell pair index uv. If that is too fine grained (though I suspect it won't be simply because each uv loop also would loop over the auxiliary shells), it's easy enough to simply set a batch size in each OpenMP parallel for pragma.

Reply:
The reason why we do that is to avoid race conditions.

The tensor contractions:
`gamma_P += (P|uv)*D_uv` (would need to block over P to avoid race conditions)
`J_uv += `(uv|Q)*gamma_Q` (would need to block over uv to avoid race conditions)